### PR TITLE
Weppy 0.9.0

### PR DESCRIPTION
**Pull request by wladekb**
Weppy 0.9.0

Replacement for Bucky at Wikia

/cc @macbre @rogatty

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+Example code
+------------
+
+You can get the following data reported to the server:
+```json
+{
+	"context": {
+		"url": "http://my.domain.com/article/Some_great_article",
+		"skin": "modern"
```
Comment (macbre) [inline README.md]: Typo? `,` is missing

File: README.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+		"user-agent": "Whatever User Agent the visitor has"
+	},
+	"data": {
+		"Lightbox.open": [
+			[ 0.24, { "image-count": 12 } ],
+			[ 0.12, { "image-count": 3 } ]
+		],
+		"pageview.fromuser.Testuser": {
+			[ 1 ]
```
Comment (macbre) [inline README.md:29]: Add a key

File: README.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
+Weppy
+=====
+
+Weppy is a client-side library for reporting performance data. It can automatically collect page load data
+using window.performance API and allows any other metric to be reported as well.
+
+Weppy is heavily inspired by [BuckyClient](https://github.com/HubSpot/BuckyClient). However it puts more focus into
+structure of the reported data, allows custom attributes to be reported with the measurement itself and supports
+sending the page context with all the measurements done on a page. See (comparison with Bucky)[#Comparison with Bucky]
```
Comment (rogatty) [inline README.md]: `[comparison with Bucky](#Comparison with Bucky)`

File: src/weppy.ts [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+                    time = timing[key];
+                    if (time && typeof time == 'number') {
+                        data[key] = (time - start);
+                    }
+                }
+            }
+            delete data['navigationStart'];
+
+            var name = '_'; // @todo: think through
+            self.namespace('PAGELOAD').timer.send(name,start,data);
```
Comment (macbre) [inline src/weppy.ts]: What is the purpose of `_` here? :)

**Comment by macbre** [comment]
LGTM

File: gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+all_deps.push('copy-weppy.d.ts');
+
+gulp.task('test',function(){
+	return gulp.src('./tests/index.html')
+		.pipe(qunit());
+});
+
+gulp.task('default', all_deps);
+
+
```
Comment (rogatty) [inline gulpfile.js]: Too many `n`

File: src/weppy.ts [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+        }
+        annotate(annotations) {
+            this._timer.annotate(this.name,annotations);
+        }
+    }
+
+    updateActive();
+}
+
+var Weppy : WeppyNamespace = new WeppyImpl.Namespace('','');
```
Comment (rogatty) [inline src/weppy.ts]: Missing `\n`

File: src/weppy.ts [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+interface Window {
+    XMLHttpRequest?: XMLHttpRequest;
+    XDomainRequest?: XMLHttpRequest;
+    JSON?: JSON;
+}
+
+module WeppyImpl {
+    var PROTOCOL_VERSION = 3;
+
+    var PATH_DELIMITER = '.';
```
Comment (rogatty) [inline src/weppy.ts]: https://github.com/Wikia/guidelines/blob/master/JavaScript/CodingConventions.md#variables

> Declare all variables using one var keyword at the top of their scope context.

**Comment by rogatty** [comment]
@wladekb can you run autoformatting on `src/weppy.ts` and `tests/weppy-test.js` files?

File: src/weppy.ts [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+    var NAMESPACE_DELIMITER = '::';
+    var active = false;
+    var options : WeppySettings = {
+        "host": '/weppy',
```
Comment (rogatty) [inline src/weppy.ts]: Can you stick to single quotes in every js/ts file?

File: tests/weppy-test.js [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+		options = {
+			host: '/weppy',
+			sample: 100,
+			active: 1,
+			aggregationInterval: 0,
+			maxInterval: 0,
+			decimalPrecision: 3,
+			transport: function( data ) {
+				if (test.check) {
+					if (test.check(data) !== false) {
```
Comment (rogatty) [inline tests/weppy-test.js]: This can be merged to one condition.

File: src/weppy.ts [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+module WeppyImpl {
+
+	var PROTOCOL_VERSION = 3,
+		PATH_DELIMITER = '.',
+		NAMESPACE_DELIMITER = '::',
+		active = false,
+		options:WeppySettings = {
+			"host": '/weppy',
+			"transport": 'url', // 'url' or 'post'
+			"active": false,
```
Comment (rogatty) [inline src/weppy.ts]: In `README.md` it's described as `true` by default.

Comment (wladekb) [inline src/weppy.ts]: Thanks for spotting this. Fixed.

**Comment by kenkouot** [comment]
Thanks for following the [JS Packages guidelines](https://github.com/Wikia/guidelines/blob/master/JavaScript/Packages.md) :+1:

File: src/weppy.ts [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+			"context": {},
+			"debug": false
+		},
+		initTime = +(new Date),
+		queue, aggregationTimeout, maxTimeout, sentPerformanceData,
+		now = () => {
+			return window.performance && window.performance.now ? window.performance.now() : +(new Date);
+		},
+		log:any = () => {
+			if (options.debug)
```
Comment (kenkouot) [inline src/weppy.ts]: Don't omit curly braces

File: src/weppy.ts [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+				if (typeof options.debug == 'function')
+					options.debug.apply(window, arguments);
+				else
+					window.console && window.console.log && window.console.log.apply
+						? window.console.log.apply(window.console, arguments) : void 0;
+		},
+		logError:any = () => {
```
Comment (kenkouot) [inline src/weppy.ts]: space after `:`

File: src/weppy.ts [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+			window.console && window.console.error && window.console.error.apply &&
+				window.console.error.apply(window.console,arguments);
+		};
+
+	function round(num:number, precision = options.decimalPrecision) {
```
Comment (kenkouot) [inline src/weppy.ts]: declare explicit return type

File: src/weppy.ts [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+		return Math.round(num * Math.pow(10, precision)) / Math.pow(10, precision)
+	}
+
+	function buildPath(path:string, subpath:string, glue:string = PATH_DELIMITER) {
+		return path + (path != '' && subpath != '' ? glue : '' ) + subpath;
```
Comment (kenkouot) [inline src/weppy.ts]: wrap the condition in the ternary operator for ease of reading and also use strict operator `!==`

File: src/weppy.ts [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+		logError:any = () => {
+			window.console && window.console.error && window.console.error.apply &&
+				window.console.error.apply(window.console,arguments);
+		};
+
+	function round(num:number, precision = options.decimalPrecision) {
+		return Math.round(num * Math.pow(10, precision)) / Math.pow(10, precision)
+	}
+
+	function buildPath(path:string, subpath:string, glue:string = PATH_DELIMITER) {
```
Comment (kenkouot) [inline src/weppy.ts]: Style guidelines inherited from JS, needs ws after function name, `function buildPath (`

File: src/weppy.ts [hunk 16]
```diff
@@ -0,0 +1,469 @@
...
+// these seem not to be defined in window
+interface Window {
+	XMLHttpRequest?: XMLHttpRequest;
+	XDomainRequest?: XMLHttpRequest;
+	JSON?: JSON;
+}
+
+module WeppyImpl {
+
+	var PROTOCOL_VERSION = 3,
```
Comment (kenkouot) [inline src/weppy.ts]: Our guidelines for consts: https://github.com/Wikia/guidelines/blob/master/JavaScript/CodingConventions.md#constants
TLDR, no special casing for consts

File: src/weppy.ts [hunk 17]
```diff
@@ -0,0 +1,469 @@
...
+			[key: number]: Node;
+		}
+		export interface Data {
+			value: number;
+			count?: number;
+			rollingAverage?: boolean;
+		}
+	}
+	class Queue {
+		private all:QueueData.Root;
```
Comment (kenkouot) [inline src/weppy.ts]: space after `:`

File: src/weppy.ts [hunk 18]
```diff
@@ -0,0 +1,469 @@
...
+				scope = this.all[name] = this.all[name] || {}, data:QueueData.Data;
+			if (serializedAnnotations) {
+				scope = scope.annotated = scope.annotated || {};
+				scope = scope[serializedAnnotations] = scope[serializedAnnotations] || {};
+			}
+			data = scope.data = scope.data || {value: 0};
+			return data;
+		}
+
+		get_clear() {
```
Comment (kenkouot) [inline src/weppy.ts]: camelCase

File: src/weppy.ts [hunk 19]
```diff
@@ -0,0 +1,469 @@
...
+			var measurements = {};
+
+			function addMeasurement(name, data, annotations) {
+				if (data) {
+					var value = data.value;
```
Comment (kenkouot) [inline src/weppy.ts]: single var statement per function, at top of function

File: src/weppy.ts [hunk 20]
```diff
@@ -0,0 +1,469 @@
...
+					if (data.rollingAverage) {
+						value = round(value);
+					}
+					var measurement = [value];
```
Comment (kenkouot) [inline src/weppy.ts]: single var statement per function, at top of function

File: src/weppy.ts [hunk 21]
```diff
@@ -0,0 +1,469 @@
...
+		return callWrapNamespace ?
+			() => {
+				return wrapNamespaceObject(fn.apply(scope,arguments));
+			} :
+			() => {
+				return fn.apply(scope,arguments);
+			}
+	}
+
+	function wrapNamespaceObject( obj: WeppyNamespace ) {
```
Comment (kenkouot) [inline src/weppy.ts]: no space around arguments inside parens, correct: `(obj: WeppyNamespace)`

File: src/weppy.ts [hunk 22]
```diff
@@ -0,0 +1,469 @@
...
+		sendData(all_data);
+	}
+
+	function sendData(data) {
+		if (typeof options.transport == 'function') {
+			options.transport(data);
+			return;
+		}
+		var url = options.host + '/v' + PROTOCOL_VERSION + '/send';
+		if (options.transport == 'url') {
```
Comment (kenkouot) [inline src/weppy.ts:240]: strict equality operator `===`

File: src/weppy.ts [hunk 23]
```diff
@@ -0,0 +1,469 @@
...
+		if (!active) {
+			log("Weppy: would send queue but inactive");
+			return;
+		}
+		if (!window.JSON || !JSON.stringify) {
+			queue.clear();
+			return;
+		}
+		var all_measurements = queue.get_clear();
+		var all_data = {
```
Comment (kenkouot) [inline src/weppy.ts]: camelCase

File: src/weppy.ts [hunk 24]
```diff
@@ -0,0 +1,478 @@
...
+				scope = this.all[name] = this.all[name] || {}, data: QueueData.Data;
+			if (serializedAnnotations) {
+				scope = scope.annotated = scope.annotated || {};
+				scope = scope[serializedAnnotations] = scope[serializedAnnotations] || {};
+			}
+			data = scope.data = scope.data || {value: 0};
+			return data;
+		}
+
+		get_clear () {
```
Comment (kenkouot) [inline src/weppy.ts]: camelCase

File: src/weppy.ts [hunk 25]
```diff
@@ -0,0 +1,478 @@
...
+	function clearSchedule () {
+		clearTimeout(aggregationTimeout);
+		clearTimeout(maxTimeout);
+		aggregationTimeout = null;
+		maxTimeout = null;
+	}
+
+
+	function sendQueue () {
+		var all_measurements, all_data;
```
Comment (kenkouot) [inline src/weppy.ts]: camelCase :)

File: src/weppy.ts [hunk 26]
```diff
@@ -0,0 +1,478 @@
...
+			"debug": false
+		},
+		initTime = +(new Date),
+		queue, aggregationTimeout, maxTimeout, sentPerformanceData,
+		now = () => {
+			return window.performance && window.performance.now ? window.performance.now() : +(new Date);
+		},
+		log: any = () => {
+			if (options.debug) {
+				if (typeof options.debug == 'function') {
```
Comment (kenkouot) [inline src/weppy.ts]: strict equality

File: src/weppy.ts [hunk 27]
```diff
@@ -0,0 +1,478 @@
...
+}
+
+module WeppyImpl {
+
+	var protocolVersion = 3,
+		pathDelimiter = '.',
+		namespaceDelimiter = '::',
+		active = false,
+		options: WeppySettings = {
+			"host": '/weppy',
```
Comment (kenkouot) [inline src/weppy.ts:17]: These object keys don't have to be in quotes

File: src/weppy.ts [hunk 28]
```diff
@@ -0,0 +1,478 @@
...
+		queue, aggregationTimeout, maxTimeout, sentPerformanceData,
+		now = () => {
+			return window.performance && window.performance.now ? window.performance.now() : +(new Date);
+		},
+		log: any = () => {
+			if (options.debug) {
+				if (typeof options.debug == 'function') {
+					options.debug.apply(window, arguments);
+				} else {
+					window.console && window.console.log && window.console.log.apply
```
Comment (kenkouot) [inline src/weppy.ts:38]: We don't support any browsers that don't support console.log, additionally, console.log fallbacks exist in many other libraries and codebases, so no need to check for it here.

File: src/weppy.ts [hunk 29]
```diff
@@ -0,0 +1,478 @@
...
+						? window.console.log.apply(window.console, arguments) : void 0;
+				}
+			}
+		},
+		logError: any = () => {
+			window.console && window.console.error && window.console.error.apply &&
```
Comment (kenkouot) [inline src/weppy.ts:44]: see above re: console.log check

File: src/weppy.ts [hunk 30]
```diff
@@ -0,0 +1,478 @@
...
+			[key: string]: Node;
+			[key: number]: Node;
+		}
+		export interface Data {
+			value: number;
+			count?: number;
+			rollingAverage?: boolean;
+		}
+	}
+	class Queue {
```
Comment (kenkouot) [inline src/weppy.ts:104]: Does the Queue structure have to live in this file?

File: src/weppy.ts (hunk 27 above)
Comment (wladekb) [inline src/weppy.ts:17]: I specifically put them in quotes to prevent name mangling when using google closure compiler as they need to stay as they are. However it seems we won't use it for minification. Do you think they should stay quoted or not in that case?

File: src/weppy.ts (hunk 30 above)
Comment (wladekb) [inline src/weppy.ts:104]: I think since the full code base is just 500 lines there's no reason to split it into multiple files.

File: src/weppy.ts (hunk 29 above)
Comment (wladekb) [inline src/weppy.ts:44]: I think it doesn't hurt us and it increases open-source value for that project to check for console.log existence and according to https://developer.mozilla.org/en-US/docs/Web/API/Console.log it's a non-standard feature. I'd leave as it is.

**Comment by wladekb** [comment]
@macbre @hakubo @rogatty @kenkouot Last call for comments. I'd like to get it merged in a day or two.
